treemix analysis: done on the core genome (obtained via snippy)  
tool into account the linkage via the treemix parameters (option k, with blocks of 10000 snps to be sure that they are unlinked)  
consider the whole core genome (coding and non coding sequence)  
however, need the snps to be biallelic  

The scripts plink2treemix.py, vcf2treemix, plotting_funcs.R have been downloaded from https://speciationgenomics.github.io/Treemix/


REQUIREMENTS  
Treemix v1.12  
snippy
vcftools v0.1.16  
python v2.7.18 (for plink2treemix.py, libraries: sys, os, gzip) 
python v3
R v4.1.1 (RColorBrewer v1.1-2, R.utils v2.11.0, OptM)

for the following, use the file core.vcf generated by snippy  
  

additional input files  

input.txt = tab separated file, one line per sequence  
-> ID	path/seq.fasta

core.clust = tab separated file  
-> sample.name	sample.name	pop/group  


get the samples that are in core.noN.vcf.gz and core.clust
> prepData.r

-> on the server (did not do this step, directly used the core.vcf previously done by Kaisa/Harry)  
```
snippy-multi input.txt --ref REF.gbk --cpu 20 > snippy_multi.sh
sh snippy_multi.sh > outSnippy.out

gzip -c core.vcf > core.vcf.gz
```

-> on my computer  (python3 installed)
```
conda activate WORK
vcftools --gzvcf core.vcf.gz --max-missing 1 --recode --stdout | gzip > core.noN.vcf.gz
bash vcf2treemix.sh core.noN.vcf.gz core.clust
```
-> note that it will work (in my case) only up to the call to the script plink2treemix.py   
-> as the script was written for python2   
```
conda activate PYTHON2
gzip core.noN.frq.strat 
python2 plink2treemix.py "core.noN.frq.strat.gz" "core.noN.treemix.frq.gz"
conda deactivate
```

-> run the part after the call to the python script from the bash script vcf2treemix.sh (by end, copy paste it)  
-> do not forget to generate the variable $file  

-> i = nbr of migration edges  
```
for i in 0 1 2 3 4 5 6 7 8 9 10; do
    treemix -i core.noN.treemix.frq.gz -m $i -o core.noN.$i -root hpAfrica2 -bootstrap -k 10000 > treemix_${i}_log &
done

for m in {0..15}; do
    for i in {1..10}; do
        # Generate random seed
        s=$RANDOM
        treemix -i core.noN.treemix.frq.gz -o core.noN.${m}.${i} -root hpAfrica2 -m ${m} -bootstrap -k 10000 -seed ${s} >> treemix_${m}.log &
    done
done
```


calculate the f3-statistics (admixture)  
```
threepop -i core.noN.treemix.frq.gz -k 500 > threepop.out
```

calculate f4-stat (gene flow)  
```
fourpop -i core.noN.treemix.frq.gz -k 500 > fourpop.out
```


scripts to plot the admixture tree
> plotting_funcs.R

and the treemix results in general
> plotRes.r
